cbits/cbits.c

```diff
@@ -237,29 +276,36 @@ _hs_text_encode_utf8(uint8_t **destp, const uint16_t *src, size_t srcoff,
 
  ascii:
 #if defined(__x86_64__)
-  while (srcend - src >= 4) {
-    uint64_t w = *((uint64_t *) src);
+  while (srcend - src >= 8) {
+    union { uint64_t halves[2]; __m128i whole; } eight_chars;
+    eight_chars.whole = _mm_loadu_si128((__m128i *) src);
 
+    const uint64_t w = eight_chars.halves[0];
     if (w & 0xFF80FF80FF80FF80ULL) {
```

Comment:
Could we just `break` here actually? Does it affect performance? I imagine we could compare a whole 128-byte register with a broadcasted `0xFF80`, and either `break` or `_mm_packus_epi16` + `_mm_storel_epi64`.

Reply:
Yeah, I experimented with where exactly and after how many checks the break can be and chose one of the best performing combinations. Unfortunately I didn't record those combinations and the results might be different on different CPUs.

Reply:
I mean, it looks asymmetric to exercise each pair of bytes from the first `uint64`, but not in the second one. I understand that doing the same for the second `uint64` make code more hairy, so maybe we can stop doing it for the first one as well? It would also allow us to avoid `union` stuff.

Reply:
I tried both symmetric variants and found both to be slower than this one.

My understanding is that

1. if you do too many checks by looking at individual bytes, the time cost of the simd loop iteration becomes higher.
2. if you do too few checks (by not even looking at bytes in the first half) then the probability of the simd loop iteration doing useful work decreases and it also hurts the overall performance.
 
This probability is different for different input data, extreme cases would be pure ASCII (simd routine always works) and pure chinese text (simd routine never works). On our test data I observed that inspecting bytes in the first half was a sweet spot.